tools/shield_utils.py

```diff
@@ -0,0 +1,427 @@
+
+    def func(acc, cpu):
+        if _is_range(cpu):
+            acc.update(_range(cpu))
+        else:
+            acc.add(int(cpu))
+        return acc
+
+    return list(functools.reduce(func, groups, arr))
+
+
+def _is_node_online(numa_node_id):
+    """Check whether a NUMA node is online.
+
+    Args:
+        * numa_node_id (int).
+    Returns:
+        * boolean.
+    """
+    online_numa_nodes_path = SYSFS_MOUNTPOINT + "/devices/system/node/online"
+    with open(online_numa_nodes_path) as fp:
+        content = fp.readline()
+        online_numa_nodes = _parse_list_format(content)
+        return len(list(filter(lambda cpu: cpu == numa_node_id,
+                               online_numa_nodes))) > 0
+
+
+def _get_cpulist(numa_node_id, raw=False):
+    """Get the cpulist of a NUMA node.
+
+    Args:
+        * numa_node_id (int).
+        * raw (boolean): Default is False. If set to True, will return
+                         the content found in the cpulist information
+                         present in sysfs.
+     Returns:
+        * str: the content found in the NUMA node cpulist file.
+        * list of ints: ints corresponding to elements of a LIST FORMAT, as
+                        described into `cpuset` documentation.
+    """
+    cpulist_path = SYSFS_MOUNTPOINT + "/devices/system/node/node{}/cpulist"\
+        .format(numa_node_id)
+    with open(cpulist_path) as fp:
+        if raw:
+            return fp.readline().strip()
+        return _parse_list_format(fp.readline())
+
+
+def _is_memory_node_available(numa_node_id):
+    """Check if the memory node is used by an explicit cpuset.
+
+    By explicit cpusets we mean cpusets different than root and '/docker'
+
+    Args:
+        * numa_node_id (int).
+    Returns:
+        * boolean.
+    """
+    (root, dirnames, _) = next(os.walk(CPUSET_MOUNTPOINT))
+    for dirname in dirnames:
+        if dirname == "docker":
+            continue
+
+        with open(root + "/" + dirname + "/cpuset.mems") as fp:
+            log.say(root + "/" + dirname + "/cpuset.mems")
+            mems = _parse_list_format(fp.readline())
+            if not len(list(filter(lambda m: m == numa_node_id, mems))) == 0:
+                return False
+
+    return True
+
+
+def _get_non_available_cpus():
+    """Return a set non available CPU cores.
+
+    A non available CPU core is a cpu id which is used by
+    others cpuset than `/` (root) and `/docker`.
+
+    Returns:
+        * list of ints: a list of CPU core IDs used by other cpusets than `/`
+                        (root) and `/docker`.
+    """
+    (root, dirnames, _) = next(os.walk(CPUSET_MOUNTPOINT))
+    non_avail_cpus = set()
+    for dirname in dirnames:
+        if dirname == "docker":
+            continue
+
+        with open(root + "/" + dirname + "/cpuset.cpus") as fp:
+            non_avail_cpus |= set(_parse_list_format(fp.readline()))
+
+    return list(non_avail_cpus)
+
+
+def _get_available_cpus(numa_node_id):
```

Comment:
This logic is quite complicated, but it's also not bulletproof. A cpu may be available when the test starts, and then it can be added to another cpuset, which would affect the results of the test. It would be much easier and reliable if we used the `cpuset.cpu_exclusive` flag in order to mark the cpus that we use as exclusive to our cpuset. WDYT ?

Reply:
Good idea. Will try this.